projects/akita-ng-fire/src/lib/collection/collection.service.ts

```diff
@@ -497,74 +496,58 @@ export class CollectionService<S extends EntityState<any, string>>  {
-    stateFnOrWrite?: UpdateStateCallback<getEntityType<S>> | Partial<getEntityType<S>> | WriteOptions,
+    idsOrEntity: Partial<EntityType> | Partial<EntityType>[] | string | string[],
+    stateFnOrWrite?: UpdateStateCallback<EntityType> | Partial<EntityType> | WriteOptions,
     options: WriteOptions = {}
   ): Promise<void | firestore.Transaction[]> {
 
     let ids: string[] = [];
-    let newStateOrFn = stateFnOrWrite as UpdateStateCallback<getEntityType<S>> | Partial<getEntityType<S>>;
+    let stateFunction: UpdateStateCallback<EntityType>;
+    let getData: (docId: string) => Partial<EntityType>;
 
-    const isEntity = (value): value is Partial<getEntityType<S>> => {
+    const isEntity = (value): value is Partial<EntityType> => {
       return typeof value === 'object' && value[this.idKey];
     };
-    const isEntityArray = (values): values is Partial<getEntityType<S>>[] => {
+    const isEntityArray = (values): values is Partial<EntityType>[] => {
       return Array.isArray(values) && values.every(value => isEntity(value));
     };
 
-    const { ctx } = options;
-    const path = this.getPath(options);
-
-    // If this is a list of entities
-    if (isEntityArray(idsOrEntity)) {
-      const { write = this.db.firestore.batch() } = options;
-      const operations = (idsOrEntity as Partial<getEntityType<S>>[]).map(async entity => {
-        const doc = Object.freeze(entity);
-        const data = this.preFormat(doc);
-        if (!entity[this.idKey]) {
-          throw new Error(`Document should have an unique id to be updated, but none was found in ${entity}`);
-        }
-        const { ref } = this.db.doc(`${path}/${entity[this.idKey]}`);
-        write.update(ref, this.formatToFirestore (data));
-        if (this.onUpdate) {
-          await this.onUpdate(data, { write, ctx });
-        }
-      });
-      await Promise.all(operations);
-      // If there is no atomic write provided
-      if (!options.write) {
-        return (write as firestore.WriteBatch).commit();
-      }
-      return;
-    }
-
     if (isEntity(idsOrEntity)) {
       ids = [ idsOrEntity[this.idKey] ];
-      options = stateFnOrWrite as WriteOptions || {} ;
-      newStateOrFn = idsOrEntity;
-    } else {
+      getData = () => idsOrEntity;
+      options = stateFnOrWrite as WriteOptions || {};
+    } else if (isEntityArray(idsOrEntity)) {
+      let entityMap = new Map(idsOrEntity.map(entity => [entity[this.idKey] as string, entity]));
+      ids = Array.from(entityMap.keys());
+      getData = docId => entityMap.get(docId);
+      options = stateFnOrWrite as WriteOptions || {};
+    } else if (typeof stateFnOrWrite === 'function') {
       ids = Array.isArray(idsOrEntity) ? idsOrEntity : [idsOrEntity];
+      stateFunction = stateFnOrWrite as UpdateStateCallback<EntityType>;
+    } else if (typeof stateFnOrWrite === 'object' && typeof idsOrEntity === 'string') {
+      ids = [idsOrEntity];
```

Comment:
should be `ids = Array.isArray(idsOrEntity) ? idsOrEntity : [idsOrEntity]`

Reply:
I did it intentionally, cause in case stateFnOrWrite, it makes no sense updating more than one document with same change... or maybe it does?

Reply:
ok, never mind, didn't notice your comment under update signature